Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -185,102 +201,68 @@ public final class KafkaConsumer {
         partition: KafkaPartition,
         offset: Int
     ) throws {
-        assert(!self.closed)
-
-        guard let partitionPointer = rd_kafka_topic_partition_list_add(
-            self.subscribedTopicsPointer,
-            topic,
-            partition.rawValue
-        ) else {
-            fatalError("rd_kafka_topic_partition_list_add returned invalid pointer")
-        }
+        let action = self.stateMachine.withLockedValue { $0.setUpConnection() }
+        switch action {
+        case .setUpConnection(let client, let subscribedTopicsPointer):
+            guard let partitionPointer = rd_kafka_topic_partition_list_add(
+                subscribedTopicsPointer,
+                topic,
+                partition.rawValue
+            ) else {
+                fatalError("rd_kafka_topic_partition_list_add returned invalid pointer")
+            }
 
-        partitionPointer.pointee.offset = Int64(offset)
+            partitionPointer.pointee.offset = Int64(offset)
 
-        let result = self.client.withKafkaHandlePointer { handle in
-            rd_kafka_assign(handle, self.subscribedTopicsPointer)
-        }
+            let result = client.withKafkaHandlePointer { handle in
+                rd_kafka_assign(handle, subscribedTopicsPointer)
+            }
 
-        guard result == RD_KAFKA_RESP_ERR_NO_ERROR else {
-            throw KafkaError.rdKafkaError(wrapping: result)
+            guard result == RD_KAFKA_RESP_ERR_NO_ERROR else {
+                throw KafkaError.rdKafkaError(wrapping: result)
+            }
         }
     }
 
-    /// Receive new messages and forward the result to the ``messages`` `AsyncSequence`.
-    func produceMore() {
-        self.serialQueue.async {
-            guard !self.closed else {
-                return
-            }
-
-            let messageResult: Element
-            do {
-                guard let message = try self.poll() else {
-                    self.produceMore()
-                    return
+    /// Start polling Kafka for messages.
+    ///
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    public func run() async throws {
+        // TODO(felix): make pollInterval part of config -> easier to adapt to Service protocol (service-lifecycle)
+        while !Task.isCancelled {
+            let nextAction = self.stateMachine.withLockedValue { $0.nextPollLoopAction() }
+            switch nextAction {
+            case .pollForAndYieldMessage(let pollInterval, let client, let source, let logger):
+                let messageResult: Result<KafkaConsumerMessage, KafkaError>
+                do {
+                    guard let message = try client.consumerPoll() else {
+                        break
+                    }
+                    messageResult = .success(message)
+                } catch let kafkaError as KafkaError {
+                    messageResult = .failure(kafkaError)
```

Comment:
If we fail polling shouldn't we shutdown everything?

Reply:
So `consumerPoll` `throws` when there is an error with the received message. However, I think we shouldn't close the entire consumer because of, e.g. a message with a faulty payload.

Reply:
Hm, I think this is pretty bad. If we get a bad message something seriously went wrong. At least that is what I expect. Can you investigate when this could happen and document it here. IMO it is weird that our sequence contains a `Result<>`, I would like to get rid of that.